ballerina/memory.bal

```diff
@@ -75,18 +155,60 @@ public isolated class MessageWindowChatMemory {
         readonly & MemoryChatMessage newMessage = check self.mapToMemoryChatMessage(message);
         lock {
             self.createSessionIfNotExist(sessionId);
-            MemoryChatMessage[] memory = self.sessions.get(sessionId);
-            if memory.length() >= self.size - 1 {
-                _ = memory.shift();
-            }
             if newMessage is MemoryChatSystemMessage {
                 self.systemMessageSessions[sessionId] = newMessage;
                 return;
             }
+
+            MemoryChatMessage[] memory = self.sessions.get(sessionId);
+            int memoryLength = memory.length();
+            ROLE newMessageRole = newMessage.role;
+            int size = self.size;
+            SummarizeOverflowConfig? config = self.summarizeOverflowConfig;
+
+            if memoryLength >= size - 1 {
+                if config is () || size <= 3 {
+                    // No summarization config or too small window - just remove oldest
+                    _ = memory.shift();
+                } else {
+                    ROLE lastMessageRole = memory[memoryLength - 1].role;
+                    boolean isLastInteractionFromAgent = lastMessageRole != USER && newMessageRole == USER;
+                    MemoryChatMessage[] slicedMemory = isLastInteractionFromAgent
+                        ? memory.slice(0, memoryLength - 1)
+                        : memory;
+
+                    MemoryChatMessage|Error summaryMessage = self.generateSummary(
+                        slicedMemory, config.modelProvider, config.summarizationPrompt, config.maxSummaryTokens);
+                    if summaryMessage is Error {
+                        // Fallback: remove oldest message if summarization fails
+                        _ = memory.shift();
+                    } else {
+                        if isLastInteractionFromAgent {
```

Comment:
In here if the last user message is from Agent, we keep the last agent mssage and summarize the rest of the buffer.
This is to keep the user,agent interaction alternatively

Reply:
A comment would be helpful here.

Reply:
Shall we add this comment in the code?